Analytics: Do Not Track disables ad-tracking and Google Analytics

MicBosi opened the pull request:
This PR disables ad-tracking/remarketing and Google Analytics when Do Not Track is enabled.

The idea is that enabling Do Not Track should have the same effect as disabling `ad-tracking` and Google Analytics from Calypso's config. Do Not Track does not affect Tracks events.

# Implementation - Ad Tracking

I've basically substituted all `config.isEnabled( 'ad-tracking' )` occurrences with the `isAdTrackingAllowed()` function (previously called `isRetargetAllowed()`) which centralises the logic on whether ad-tracking should be performed based on whether:

1. `'ad-tracking'` is disabled
2. `Do Not Track` is enabled
3. `document.location.href` may contain personally identifiable information

# Implementation - Google Analytics

Here as well, using `isGoogleAnalyticsEnabled`, I've centralised the check on whether GA should be disabled based on whether:

1. `config( 'google_analytics_enabled' )` is false
2. `Do Not Track` is enabled

There are two changes in behaviour when GA is disabled:

1. We don't load the GA script (previously we were loading it even if we weren't using it).
2. `analytics.ga.recordPageView/recordEvent/recordTiming` now immediately return. Previously they were initialising GA, logging debug info then check if GA was disabled, which appears both wasteful and misleading.

# Testing

## Enabling GA & ad-tracking

- In `config/development.json` enable both `google_analytics_enabled` and `ad-tracking`
- Restart Calypso
- In Chromes JS console enable all debug logs: `localStorage.setItem('debug', 'calypso:analytics:*');`
- Visiting for example http://calypso.localhost:3000 should show these entries:

  - `calypso:analytics:ga Recording Page View` (Google Analytics is working)
  - `calypso:analytics:ad-tracking Retargeting` (ad-tracking is working)

![do-not-track-1](https://cloud.githubusercontent.com/assets/10284338/25234855/a63d29b0-25e3-11e7-8c85-d38291bbb8bc.png)

Visiting a URL containing potential personally identifiable information like:

- http://calypso.localhost:3000/?email=noone

Should show no new `calypso:analytics:ad-tracking` logs in Chrome's JS console (ie ad tracking has been blocked for PII reasons) while `calypso:analytics:ga` entries should still be visible.

## Enabling "Do Not Track"

- Enable Do Not Track from Chrome's `Settings -> Show advanced settings -> Privacy -> Send a "Do Not Track" request`.
- Visiting for example http://calypso.localhost:3000 should show no entries for:

  - `calypso:analytics:ga` (Google Analytics is disabled)
  - `calypso:analytics:ad-tracking` (ad-tracking is disabled)

- Disable Do Not Track setting in Chrome
- Reload Calypso
- `calypso:analytics:ga` and `calypso:analytics:ad-tracking` should now be back

## Disabling GA & ad-tracking

- In `config/development.json` disable both `google_analytics_enabled` and `ad-tracking`
- Restart Calypso
- Visiting for example http://calypso.localhost:3000 should show no entries for:

  - `calypso:analytics:ga` (Google Analytics is disabled)
  - `calypso:analytics:ad-tracking` (ad-tracking is disabled)

![do-not-track-2](https://cloud.githubusercontent.com/assets/10284338/25235530/ab956fba-25e5-11e7-8b60-75f9b5a2ce3a.png)

MicBosi commented:
@travisw that's a great question. Thinking about it we should probably apply that test to GA as well. I'll implement that and update the PR. Thank you!

markryall commented on client/lib/analytics/ad-tracking.js:
```diff
@@ -197,13 +197,28 @@ function loadTrackingScripts( callback ) {
+	}
+
+	// 2. User has Do Not Track enabled
+	if ( '1' === navigator.doNotTrack ) {
```
this is used in a couple of places - maybe define a `doNotTrack` function in `client/lib/analytics/doNotTrack.js`

MicBosi commented:
@travisw @markryall thanks for your feedback! I've implemented a small `utils.js` module where we define the `doNotTrack()` and `isPiiUrl()` functions. 

GA is also blocked when a URL with potential PII is present.

I've also fixed a bug where we load GA, if enabled, even if the user has Do Not Track enabled at startup time (since it can change later).

@markryall I'd prefer not mixing the middleware optimization with this PR because:
1. fixing/improving how analytics is loaded/enabled/disabled etc. is not part of this PR scope.
2. I think reviewing how to switch on&off the various parts of `analytics` deserves careful review and its own PR.
3. Do Not Track is a transient setting that the user can change at any time while using Calypso.

But we should definitely reconvene on the topic and perhaps create a task for it with a bit more context on the whys and hows.
